File: sdk_go/records.go
// Package allyourbase.
package allyourbase

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

type ListParams struct {
	Page      int
	PerPage   int
	Sort      string
	Filter    string
	Search    string
	Fields    string
	Expand    string
	SkipTotal bool

	Fuzzy         bool
	TypoThreshold *float64
	Highlight     bool
	Facets        []string
	Semantic      bool
	SemanticQuery string
}

type GetParams struct {
	Fields string
	Expand string
}

type RecordsClient struct {
	client *Client
}

func recordsCollectionPath(collection string) string {
	return "/api/collections/" + url.PathEscape(collection)
}

func recordItemPath(collection, id string) string {
	return recordsCollectionPath(collection) + "/" + url.PathEscape(id)
}

// List returns a paginated list of records from the given collection, filtered and sorted by the provided parameters.
func (r *RecordsClient) List(ctx context.Context, collection string, params ListParams) (*ListResponse, error) {
	q := url.Values{}
	if params.Page > 0 {
		q.Set("page", strconv.Itoa(params.Page))
	}
	if params.PerPage > 0 {
		q.Set("perPage", strconv.Itoa(params.PerPage))
	}
	if params.Sort != "" {
		q.Set("sort", params.Sort)
	}
	if params.Filter != "" {
		q.Set("filter", params.Filter)
	}
	if params.Search != "" {
		q.Set("search", params.Search)
	}
	if params.Fields != "" {
		q.Set("fields", params.Fields)
	}
	if params.Expand != "" {
		q.Set("expand", params.Expand)
	}
	if params.SkipTotal {
		q.Set("skipTotal", "true")
	}
	if params.Fuzzy {
		q.Set("fuzzy", "true")
	}
	if params.TypoThreshold != nil {
		q.Set("typo_threshold", strconv.FormatFloat(*params.TypoThreshold, 'f', -1, 64))
	}
	if params.Highlight {
		q.Set("highlight", "true")
	}
	if len(params.Facets) > 0 {
		q.Set("facets", strings.Join(params.Facets, ","))
	}
	if params.Semantic {
		q.Set("semantic", "true")
	}
	if params.SemanticQuery != "" {
		q.Set("semantic_query", params.SemanticQuery)
	}
	body, err := r.client.doJSON(ctx, http.MethodGet, recordsCollectionPath(collection), q, nil)
	if err != nil {
		return nil, err
	}
	var out ListResponse
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get retrieves a single record by ID from the given collection with optional field selection and relation expansion.
func (r *RecordsClient) Get(ctx context.Context, collection, id string, params GetParams) (map[string]any, error) {
	q := url.Values{}
	if params.Fields != "" {
		q.Set("fields", params.Fields)
	}
	if params.Expand != "" {
		q.Set("expand", params.Expand)
	}
	body, err := r.client.doJSON(ctx, http.MethodGet, recordItemPath(collection, id), q, nil)
	if err != nil {
		return nil, err
	}
	var out map[string]any
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

func (r *RecordsClient) Create(ctx context.Context, collection string, data map[string]any) (map[string]any, error) {
	return r.writeRecord(ctx, http.MethodPost, recordsCollectionPath(collection), data)
}

func (r *RecordsClient) Update(ctx context.Context, collection, id string, data map[string]any) (map[string]any, error) {
	return r.writeRecord(ctx, http.MethodPatch, recordItemPath(collection, id), data)
}

func (r *RecordsClient) Delete(ctx context.Context, collection, id string) error {
	_, err := r.client.doJSON(ctx, http.MethodDelete, recordItemPath(collection, id), nil, nil)
	return err
}

func (r *RecordsClient) Batch(ctx context.Context, collection string, operations []BatchOperation) ([]BatchResult, error) {
	body, err := r.client.doJSON(ctx, http.MethodPost, recordsCollectionPath(collection)+"/batch", nil, map[string]any{"operations": operations})
	if err != nil {
		return nil, err
	}
	var out []BatchResult
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

func (r *RecordsClient) writeRecord(ctx context.Context, method, path string, data map[string]any) (map[string]any, error) {
	body, err := r.client.doJSON(ctx, method, path, nil, data)
	if err != nil {
		return nil, err
	}
	var out map[string]any
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

File: sdk/src/records.ts
/**
 * @module SDK client for managing database records via CRUD and batch operations.
 */
import type {
  BatchOperation,
  BatchResult,
  GetParams,
  ListParams,
  ListResponse,
  SearchHit,
} from "./types";
import { encodePathSegment } from "./helpers";

interface RecordsClientRuntime {
  request<T>(path: string, init?: RequestInit & { skipAuth?: boolean }): Promise<T>;
}

/**
 * HTTP client for CRUD and batch operations on database records.
 */
export class RecordsClient {
  constructor(private client: RecordsClientRuntime) {}

  /** List records in a collection with optional filtering, sorting, and pagination. */
  /**
   * Retrieves paginated records from a collection with optional filtering, sorting, search, and field selection.
   * @param collection - Collection name to query
   * @param params - Optional filtering, pagination, and response options
   * @returns Promise resolving to paginated records and metadata
   */
  async list<T = SearchHit>(
    collection: string,
    params?: ListParams,
  ): Promise<ListResponse<T>> {
    const safeCollection = encodePathSegment(collection);
    const qs = new URLSearchParams();
    if (params?.page != null) qs.set("page", String(params.page));
    if (params?.perPage != null) qs.set("perPage", String(params.perPage));
    if (params?.sort) qs.set("sort", params.sort);
    if (params?.filter) qs.set("filter", params.filter);
    if (params?.search) qs.set("search", params.search);
    if (params?.fields) qs.set("fields", params.fields);
    if (params?.expand) qs.set("expand", params.expand);
    if (params?.skipTotal) qs.set("skipTotal", "true");
    if (params?.fuzzy === true) qs.set("fuzzy", "true");
    if (params?.typoThreshold != null) qs.set("typo_threshold", String(params.typoThreshold));
    if (params?.highlight === true) qs.set("highlight", "true");
    if (params?.facets?.length) qs.set("facets", params.facets.join(","));
    if (params?.disjunctiveFacets?.length) {
      qs.set("disjunctiveFacets", params.disjunctiveFacets.join(","));
    }
    if (params?.semantic === true) qs.set("semantic", "true");
    if (params?.semanticQuery) qs.set("semantic_query", params.semanticQuery);
    if (params?.nearest) qs.set("nearest", JSON.stringify(params.nearest));
    if (params?.vectorColumn) qs.set("vector_column", params.vectorColumn);
    if (params?.distance) qs.set("distance", params.distance);
    const suffix = qs.toString() ? `?${qs}` : "";
    return this.client.request(`/api/collections/${safeCollection}${suffix}`);
  }

  /** Get a single record by primary key. */
  async get<T = Record<string, unknown>>(
    collection: string,
    id: string,
    params?: GetParams,
  ): Promise<T> {
    const safeCollection = encodePathSegment(collection);
    const safeID = encodePathSegment(id);
    const qs = new URLSearchParams();
    if (params?.fields) qs.set("fields", params.fields);
    if (params?.expand) qs.set("expand", params.expand);
    const suffix = qs.toString() ? `?${qs}` : "";
    return this.client.request(`/api/collections/${safeCollection}/${safeID}${suffix}`);
  }

  /** Create a new record. */
  async create<T = Record<string, unknown>>(
    collection: string,
    data: Record<string, unknown>,
  ): Promise<T> {
    return this.client.request(`/api/collections/${encodePathSegment(collection)}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    });
  }

  /** Update an existing record (partial update). */
  async update<T = Record<string, unknown>>(
    collection: string,
    id: string,
    data: Record<string, unknown>,
  ): Promise<T> {
    return this.client.request(
      `/api/collections/${encodePathSegment(collection)}/${encodePathSegment(id)}`,
      {
        method: "PATCH",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(data),
      },
    );
  }

  /** Delete a record by primary key. */
  async delete(collection: string, id: string): Promise<void> {
    return this.client.request(
      `/api/collections/${encodePathSegment(collection)}/${encodePathSegment(id)}`,
      {
        method: "DELETE",
      },
    );
  }

  /** Execute multiple operations in a single atomic transaction. Max 1000 operations. */
  async batch<T = Record<string, unknown>>(
    collection: string,
    operations: BatchOperation[],
  ): Promise<BatchResult<T>[]> {
    return this.client.request(`/api/collections/${encodePathSegment(collection)}/batch`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ operations }),
    });
  }
}
